Firestore/core/src/firebase/firestore/nanopb/fields_map.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_NANOPB_FIELDS_MAP_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_NANOPB_FIELDS_MAP_H_
+
+#include "Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.h"
+#include "Firestore/Protos/nanopb/firestore/local/mutation.nanopb.h"
+#include "Firestore/Protos/nanopb/firestore/local/target.nanopb.h"
+#include "Firestore/Protos/nanopb/google/firestore/v1/document.nanopb.h"
+#include "Firestore/Protos/nanopb/google/firestore/v1/firestore.nanopb.h"
+#include "Firestore/Protos/nanopb/google/type/latlng.nanopb.h"
+
+namespace firebase {
+namespace firestore {
+namespace nanopb {
+
+/**
+ * Returns a pointer to the Nanopb-generated array that describes the fields
+ * of the Nanopb proto; the array is required to call most Nanopb functions.
+ *
+ * There is always a one-to-one correspondence between a Nanopb-generated
+ * message type and its fields descriptor; essentially, the fields descriptor is
+ * a property of the type.
+ */
+// The non-specialized version of this function is deleted to make sure that
```

Comment:
One of the reasons I chose specialized template over a set of overloaded functions is that I think the error message becomes slightly clearer (otherwise, it will be diluted by the compiler listing all of the overloads that don't match). Also, this approach is more similar to a traits class, which it essentially is.